File: algorithms-and-data-structures/07. DFS, MST/I.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>

using namespace std;

vector<pair<int, int>> V;

double distance(pair<int, int> &x, pair<int, int> &y) {
    int x1 = x.first;
    int y1 = x.second;
    int x2 = y.first;
    int y2 = y.second;

    return sqrt(((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1)));
}

int main() {
    int n;
    cin >> n;

    V.resize(n, pair<int, int>(0, 0));
    vector<double> d(n, 0.0);
    vector<bool> inA(n, false);

    for (int i = 0; i < n; i++) {
        int x, y;
        cin >> x >> y;

        V[i] = pair<int, int>(x, y);
    }

    inA[0] = true;
    d[0] = -1;

    for (int i = 1; i < n; i++) {
        d[i] = distance(V[0], V[i]);
    }

    double summaryWeight = 0.0;

    for (int i = 0; i < n - 1; i++) {
        int v = 0;

        while (d[v] <= 0) {
            v++;
        }

        for (int j = v + 1; j < n; j++) {
            if (!inA[j] && d[j] < d[v]) {
                 v = j;
            }
        }

        summaryWeight += d[v];

        inA[v] = true;
        d[v] = -1.0;

        for (int u = 0; u < n; u++) {
            if (!inA[u] && distance(V[min(u, v)], V[max(u, v)]) < d[u]) {
                d[u] = distance(V[min(u, v)], V[max(u, v)]);
            }
        }
    }

    const auto digits = numeric_limits<double>::digits10;
    cout << setprecision(digits) << summaryWeight << endl;

    return 0;
}

File: algorithms-and-data-structures/07. DFS, MST/I.kt
import java.util.*
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

private lateinit var V: Array<Pair<Int, Int>>

// TL на 20м тесте, переписал на C++
fun main() {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextInt()

    V = Array(n) { Pair(0, 0) }
    val d = Array(n) { 0.0 }
    val inA = Array(n) { false }

    for (i in 0 until n) {
        val x = scanner.nextInt()
        val y = scanner.nextInt()

        V[i] = Pair(x, y)
    }

    inA[0] = true
    d[0] = -1.0

    for (i in 1 until n) {
        d[i] = distance(Pair(0, i))
    }

    var summaryWeight = 0.0

    repeat(n - 1) {
        var v = d.indexOfFirst { it > 0 }

        for (i in (v + 1) until n) {
            if (!inA[i] && d[i] < d[v]) {
                v = i
            }
        }

        summaryWeight += d[v]

        inA[v] = true
        d[v] = -1.0

        for (u in 0 until n) {
            if (!inA[u] && distance(Pair(min(u, v), max(u, v))) < d[u]) {
                d[u] = distance(Pair(min(u, v), max(u, v)))
            }
        }

    }

    println(summaryWeight)
}

fun distance(pair: Pair<Int, Int>): Double {
    return distance(V[pair.first], V[pair.second])
}

fun distance(x: Pair<Int, Int>, y: Pair<Int, Int>): Double {
    val x1 = x.first
    val y1 = x.second
    val x2 = y.first
    val y2 = y.second

    return sqrt(((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1)).toDouble())
}